### PR TITLE
[WIP] AnimateDiff Pipeline

### PR DESCRIPTION
**Pull request by Pie31415**
Issue #4524

### CODE REVIEW

File: src/diffusers/pipelines/animate_diff/motion_module.py [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+        self.video_length = video_length
+
+    def forward(self, hidden_states, temb, encoder_hidden_states, attention_mask=None, anchor_frame_idx=None):
+        hidden_states = self.temporal_transformer(hidden_states, encoder_hidden_states, attention_mask, self.video_length)
+
+        output = hidden_states
+        return output
+
+
+class TemporalTransformer3DModel(nn.Module):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/animate_diff/motion_module.py:90]: This looks to be the same as: https://github.com/huggingface/diffusers/blob/705c592ea98ba4e288d837b9cba2767623c78603/src/diffusers/models/transformer_temporal.py#L39 - can we try to use existing classes instead?

File: src/diffusers/pipelines/animate_diff/motion_module.py [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+import torchvision
+
+from ...configuration_utils import ConfigMixin, register_to_config
+from ...models.modeling_utils import ModelMixin
+from ...utils import BaseOutput
+from ...utils.import_utils import is_xformers_available
+from ...models.attention_processor import Attention
+from ...models.attention import FeedForward
+
+from einops import rearrange, repeat
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/animate_diff/motion_module.py:17]:
```suggestion
```

No einops in `diffusers` please . We don't intend to add `einops` as a dependency

File: src/diffusers/pipelines/animate_diff/motion_module.py [hunk 3]
```diff
@@ -0,0 +1,557 @@
+from dataclasses import dataclass
+from typing import List, Optional, Tuple, Union
+
+import torch
+import numpy as np
+import torch.nn.functional as F
+from torch import nn
+import torchvision
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/animate_diff/motion_module.py:8]:
```suggestion
```

`torchvision` is and should not become a dependency

File: src/diffusers/pipelines/animate_diff/motion_module.py [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+        pe[0, :, 0::2] = torch.sin(position * div_term)
+        pe[0, :, 1::2] = torch.cos(position * div_term)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x):
+        x = x + self.pe[:, :x.size(1)]
+        return self.dropout(x)
+
+
+class CrossAttention(nn.Module):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/animate_diff/motion_module.py:249]: Can we remove this class and instead use: https://github.com/huggingface/diffusers/blob/0bd5271ba832227a3170e093a8dd801742d2714e/src/diffusers/models/attention_processor.py#L34 ?

File: src/diffusers/pipelines/animate_diff/pipeline_animate_diff.py [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+    DDIMScheduler,
+    DPMSolverMultistepScheduler,
+    EulerAncestralDiscreteScheduler,
+    EulerDiscreteScheduler,
+    LMSDiscreteScheduler,
+    PNDMScheduler,
+)
+from ...utils import deprecate, logging, BaseOutput
+
+from einops import rearrange
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/animate_diff/pipeline_animate_diff.py:28]:
```suggestion
```

File: src/diffusers/pipelines/animate_diff/pipeline_animate_diff.py [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+                " \n- runwayml/stable-diffusion-inpainting \n you should change 'sample_size' to 64 in the"
+                " configuration file. Please make sure to update the config accordingly as leaving `sample_size=32`"
+                " in the config might lead to incorrect results in future versions. If you have downloaded this"
+                " checkpoint from the Hugging Face Hub, it would be very nice if you could open a Pull request for"
+                " the `unet/config.json` file"
+            )
+            deprecate("sample_size<64", "1.0.0", deprecation_message, standard_warn=False)
+            new_config = dict(unet.config)
+            new_config["sample_size"] = 64
+            unet._internal_dict = FrozenDict(new_config)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/animate_diff/pipeline_animate_diff.py:107]:
```suggestion
```

Let's not copy deprecated code

File: src/diffusers/pipelines/animate_diff/pipeline_animate_diff.py [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+            return self.device
+        for module in self.unet.modules():
+            if (
+                hasattr(module, "_hf_hook")
+                and hasattr(module._hf_hook, "execution_device")
+                and module._hf_hook.execution_device is not None
+            ):
+                return torch.device(module._hf_hook.execution_device)
+        return self.device
+
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/animate_diff/pipeline_animate_diff.py:149]:
```suggestion
```

This is not needed to be written in every pipeline anymore

File: src/diffusers/pipelines/animate_diff/pipeline_animate_diff.py [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+    def _encode_prompt(self, prompt, device, num_videos_per_prompt, do_classifier_free_guidance, negative_prompt):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/animate_diff/pipeline_animate_diff.py:150]: Let's make sure `_encode_prompt` adheres to the new format: https://github.com/huggingface/diffusers/pull/4617

**Review by patrickvonplaten — COMMENTED** [review]
Nice beginning! Let's try to re-use exsiting code components as much as possible here

**Comment by Pie31415** [comment]
@patrickvonplaten Still WIP but what's the opinion on pushing new scripts `unet_3d_motion_module.py` for example. Animate Diff modifies the unet is several ways. Would it be better to keep this as a community pipeline instead?

**Comment by patrickvonplaten** [comment]
@DN6 can you try to take over this PR here and guide, help with the review?

**Comment by DN6** [comment]
Can do

**Comment by Pie31415** [comment]
@DN6 Can you take a look at the latest code? I'm not quite sure how we want to move forward with implementing the different modules that look very similar to existing diffuser classes but with a motion module appended to it. The Temporal Transformer Model they have also seems to be different and incorporates some positional encoding and versatile attention which I am not quite sure if diffusers supports.

Also, I need some help with this but how are the pipelines usually tested by developers? I seem to be running into a lot of relative import errors with using modules from the diffusers library.

**Comment by adhikjoshi** [comment]
> @DN6 Can you take a look at the latest code? I'm not quite sure how we want to move forward with implementing the different modules that look very similar to existing diffuser classes but with a motion module appended to it. The Temporal Transformer Model they have also seems to be different and incorporates some positional encoding and versatile attention which I am not quite sure if diffusers supports.
> 
> Also, I need some help with this but how are the pipelines usually tested by developers? I seem to be running into a lot of relative import errors with using modules from the diffusers library.

+1

**Pull request closed without merging**